### DISCUSSION TITLE
How to copy file to folder .output after build(2)

### DISCUSSION THREAD

**Discussion by XCLHove** [nuxt/nuxt#27254] [Questions]
### Environment

------------------------------
- Operating System: Linux
- Node Version:     v18.20.1
- Nuxt Version:     3.11.2
- CLI Version:      3.11.1
- Nitro Version:    2.9.6
- Package Manager:  npm@10.5.0
- Builder:          -
- User Config:      modules, devtools, devServer, app, vite, nitro
- Runtime Modules:  @pinia/nuxt@0.5.1, @element-plus/nuxt@1.0.8, @vueuse/nuxt@10.9.0, @nuxtjs/device@3.1.1
- Build Modules:    -
------------------------------

### Reproduction

`nuxt.config.ts`：
```ts
import topLevelAwait from "vite-plugin-top-level-await";
import copy from "rollup-plugin-copy";
export default defineNuxtConfig({
  modules: [
    "@pinia/nuxt",
    "@element-plus/nuxt",
    "@vueuse/nuxt",
    "@nuxtjs/device",
  ],
  devtools: { enabled: true },
  devServer: {
    port: 8000,
    host: "0.0.0.0",
  },
  app: {
    head: {
      title: "XNote",
      link: [{ rel: "icon", type: "image/png", href: "/xnote.png" }],
    },
  },
  vite: {
    plugins: [
      topLevelAwait({
        promiseExportName: "__tla",
        promiseImportName: (i) => `__tla_${i}`,
      }),
      copy({
        verbose: true,
        targets: [{ src: "deploy/scf_bootstrap", dest: ".output" }],
      }),
    ],
    css: {
      preprocessorOptions: {
        less: {
          additionalData: `@import "/assets/css/main.less";`,
        },
      },
    },
  },
  nitro: {
    compressPublicAssets: true,
  },
});

```
Focus:
```ts
copy({
        verbose: true,
        targets: [
          { src: "deploy/scf_bootstrap", dest: ".output", },
        ],
})
```

### Describe the bug

我想要在构建完成后复制一些文件到`.output`文件夹的根目录下，#26623 的问题和我的类似但又有所不同，因为我要复制的文件和业务代码无关，尝试使用[rollup-plugin-copy](https://www.npmjs.com/package/rollup-plugin-copy)，发现该插件复制文件的时机是在构建完静态文件之后而不是在所有文件都构建完之后，这导致文件复制到`.output`后在后续的构建操作会清空`.output`文件夹从而导致之前复制的文件被删除，大概流程：
1. 构建静态文件
2. `rollup-plugin-copy`复制文件到`.output`
3. 清空`.output`
4. 完成后续构建
---
I want to copy some files to the root directory of the `.output` folder after the build is completed. The issue is similar to #26623 but with a difference in that the files I want to copy are unrelated to the business code. I tried using [rollup-plugin-copy](https://www.npmjs.com/package/rollup-plugin-copy) and found that the plugin copies the files after building the static files, rather than after building all files. This results in the files being copied to `.output`, but subsequent build operations clear the `.output` folder, causing the previously copied files to be deleted. The general process is as follows:
1. Build static files
2. `rollup-plugin-copy` copies files to `.output`
3. Clear `.output`
4. Complete subsequent build operations

**Comment by huang-julien** [comment]
Hello :wave: are you looking for the `build:done` build-time hook ?

**Reply by XCLHove** [reply]
Can you provide me with relevant links or configurations? Thank you.

**Comment by XCLHove** [comment]
log:
```
> build
> nuxt build

Nuxt 3.11.2 with Nitro 2.9.6                                                                                  22:56:36
[22:56:38] ℹ Building client...
[22:56:39] ℹ vite v5.2.8 building for production...
[22:56:46] ℹ ✓ 1741 modules transformed.

copied:
  deploy/scf_bootstrap → .output/scf_bootstrap

...
[22:56:49] ℹ ✓ built in 10.66s
[22:56:49] ✔ Client built in 10678ms
[22:56:49] ℹ Building server...
[22:56:49] ℹ vite v5.2.8 building SSR bundle for production...
[22:56:55] ℹ ✓ 969 modules transformed.

copied:
  deploy/scf_bootstrap → .output/scf_bootstrap

...
[22:56:57] ℹ ✓ built in 8.21s
[22:56:57] ✔ Server built in 8214ms
[nitro 22:57:07] ✔ Generated public .output/public
[nitro 22:57:07] ℹ Building Nuxt Nitro server (preset: node-server)
[nitro 22:57:12] ✔ Nuxt Nitro server built
Σ Total size: 5.52 MB (1.39 MB gzip)
[nitro 22:57:13] ✔ You can preview this build using node .output/server/index.mjs

进程已结束，退出代码为 0
```

**Reply by xjccc** [reply]
You could use a node script to achive

looks like

```
// copy.js
#!/usr/bin/env zx

const sourceDir = './source'; // 源目录路径
const targetDir = './target'; // 目标目录路径

await $`cp -R ${sourceDir}/. ${targetDir}`;

// package.json

script: {
  "build": "nuxt build && node ./copy.js"
}

```

**Reply by XCLHove** [reply]
Ok, thank you.

**Comment by RSoraM** [comment]
checkout https://nuxt.com/docs/api/advanced/hooks#nuxt-hooks-build-time

Based on your description, I think you should use the `build:done` hook.

for someone who want to copy after the build command, using `vite:compiled` hook.

```typescript
import fs from 'fs-extra'
import path from 'path'

export default defineNuxtConfig({

  hooks: {
    "vite:compiled": () => {
      const output = path.join(__dirname, '.output')
      const files = ['ecosystem.config.cjs', '...']

      for (const file of files) {
        const src = path.join(__dirname, file)
        const dest = path.join(output, file)
        fs.copySync(src, dest)
      }
    }
  },

  ...

})
```

**Comment by mirages** [comment]
My problems are similar to yours, may be you can use `nitro.publicAssets` config。

`Public asset directories to serve in development and bundle in production.`

Detail to [publicAssets](https://nitro.build/config#publicassets)

```js
// nuxt.config.ts file

export default defineNuxtConfig({
  //... some other config
  nitro: {
    publicAssets: [
      {
        dir: 'local asstets path',
        baseURL: 'server serve basePath',
        maxAge: 365 * 24 * 60 * 60
      }
    ]
  }
})
```
